crates/core/src/logstore/mod.rs

```diff
@@ -25,37 +73,65 @@ use crate::kernel::Action;
 use crate::protocol::{get_last_checkpoint, ProtocolError};
 use crate::{DeltaResult, DeltaTableError};
 
-pub(crate) mod default_logstore;
-pub(crate) mod storage;
-
+pub use self::config::StorageConfig;
+pub use self::factories::{
+    logstore_factories, object_store_factories, store_for, LogStoreFactory,
+    LogStoreFactoryRegistry, ObjectStoreFactory, ObjectStoreFactoryRegistry,
+};
 pub use self::storage::utils::{commit_uri_from_version, str_is_truthy};
-#[cfg(feature = "cloud")]
-pub use self::storage::RetryConfigParse;
 pub use self::storage::{
-    factories, limit_store_handler, store_for, url_prefix_handler, DefaultObjectStoreRegistry,
-    DeltaIOStorageBackend, IORuntime, ObjectStoreFactory, ObjectStoreRef, ObjectStoreRegistry,
-    ObjectStoreRetryExt, StorageOptions,
+    DefaultObjectStoreRegistry, DeltaIOStorageBackend, IORuntime, ObjectStoreRef,
+    ObjectStoreRegistry, ObjectStoreRetryExt,
 };
+/// Convenience re-export of the object store crate
 pub use ::object_store;
 
-/// Trait for generating [LogStore] implementations
-pub trait LogStoreFactory: Send + Sync {
-    /// Create a new [LogStore]
-    fn with_options(
+mod config;
+pub(crate) mod default_logstore;
+pub(crate) mod factories;
+pub(crate) mod storage;
+
+/// Internal trait to handle object store configuration and initialization.
+trait LogStoreFactoryExt {
+    fn with_options_internal(
+        &self,
+        store: ObjectStoreRef,
+        location: &Url,
+        options: &StorageConfig,
+        io_runtime: Option<IORuntime>,
+    ) -> DeltaResult<Arc<dyn LogStore>>;
+}
+
+impl<T: LogStoreFactory + ?Sized> LogStoreFactoryExt for T {
+    fn with_options_internal(
+        &self,
+        store: ObjectStoreRef,
+        location: &Url,
+        options: &StorageConfig,
+        io_runtime: Option<IORuntime>,
+    ) -> DeltaResult<Arc<dyn LogStore>> {
+        let store = options.decorate_store(store, location, io_runtime.map(|r| r.get_handle()))?;
+        self.with_options(Arc::new(store), location, options)
+    }
+}
```

Comment:
Right now we use the internal trait to make sure we are injection ourselves into the flow to wrap our layers around the object store. I feel this can be improved, bit hoping to do that in a follou up where we can have a more targeted discussion around this.